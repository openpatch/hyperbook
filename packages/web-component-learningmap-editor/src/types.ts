import { Node, Edge, Box } from "@xyflow/react";

// ============================================================================
// TYPES & INTERFACES
// ============================================================================

export interface UnlockCondition {
  after?: string[];
  date?: string;
  password?: string;
}

export interface CompletionNeed {
  id: string;
  source?: string;
  target?: string;
}

export interface Completion {
  needs?: CompletionNeed[];
  optional?: CompletionNeed[];
}

export interface NodeData {
  state: "locked" | "unlocked" | "started" | "completed" | "mastered";
  label: string;
  description?: string;
  duration?: string;
  unlock?: UnlockCondition;
  completion?: Completion;
  video?: string;
  resources?: { label: string; url: string }[];
  summary?: string;
  [key: string]: any;
}

export interface ImageNodeData {
  data?: string; // base64 encoded image
}

export interface TextNodeData {
  text?: string;
  fontSize?: number;
  color?: string;
  rotation?: number;
}

export type BackgroundNodeData = ImageNodeData | TextNodeData;

export interface BackgroundConfig {
  color?: string;
  nodes?: Node<BackgroundNodeData>[];
}

export interface EdgeConfig {
  animated?: boolean;
  color?: string;
  width?: number;
  type?: string;
}

export interface RoadmapData {
  nodes?: Node<NodeData>[];
  edges?: Edge[];
  background?: BackgroundConfig;
  edgeConfig?: EdgeConfig;
}

export type Orientation = "horizontal" | "vertical";

export type HelperLine = {
  // Used to filter out helper lines corresponding to the node being dragged
  node: Node;
  // We use it to check that the helper line is within the viewport.
  nodeBox: Box;
  // 0 for horizontal, 1 for vertical
  orientation: Orientation;
  // If orientation is 'horizontal', `position` holds the Y coordinate of the helper line.
  // (Might correspond to the top or bottom position of a node, or other anchors).
  // If orientation is 'vertical', `position` holds the X coordinate of the helper line.
  position: number;
  // Optional color for the helper line
  color?: string;
  anchorName: string;
};
