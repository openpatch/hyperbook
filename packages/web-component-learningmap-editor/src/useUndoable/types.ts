export type ActionType =
  | "undo"
  | "redo"
  | "update"
  | "reset"
  | "resetInitialState";

export type HistoryLimit = number | "infinium" | "infinity";

export type MutationBehavior =
  | "mergePastReversed"
  | "mergePast"
  | "destroyFuture"
  | "keepFuture";

export interface Action<T> {
  type: ActionType;
  payload?: T;
  behavior?: MutationBehavior;
  historyLimit?: HistoryLimit;
  ignoreIdenticalMutations?: boolean;
  cloneState?: boolean;
  ignoreAction?: boolean;
}

export interface State<T> {
  past: T[];
  present: T;
  future: T[];
}

export interface Options {
  behavior?: MutationBehavior;
  historyLimit?: HistoryLimit;
  ignoreIdenticalMutations?: boolean;
  cloneState?: boolean;
}

export type UseUndoable<T> = [
  T,
  (
    payload: T | ((oldValue: T) => T),
    behavior?: MutationBehavior,
    ignoreAction?: boolean,
  ) => void,
  {
    past: T[];
    future: T[];

    undo: () => void;
    canUndo: boolean;
    redo: () => void;
    canRedo: boolean;

    reset: (initialState?: T) => void;
    resetInitialState: (newInitialState: T) => void;
    static_setState: (
      payload: T,
      behavior?: MutationBehavior,
      ignoreAction?: boolean,
    ) => void;
  },
];
