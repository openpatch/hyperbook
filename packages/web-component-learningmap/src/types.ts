import { Node, Edge, Box } from "@xyflow/react";

// ============================================================================
// TYPES & INTERFACES
// ============================================================================

export interface UnlockCondition {
  after?: string[];
  date?: string;
  password?: string;
}

export interface Completion {
  needs?: string[];
  optional?: string[];
}

export interface NodeData {
  state: "locked" | "unlocked" | "started" | "completed" | "mastered";
  label: string;
  description?: string;
  duration?: string;
  unlock?: UnlockCondition;
  completion?: Completion;
  video?: string;
  resources?: { label: string; url: string }[];
  summary?: string;
  [key: string]: any;
}

export interface ImageNodeData {
  data?: string; // base64 encoded image
}

export interface TextNodeData {
  text?: string;
  fontSize?: number;
  color?: string;
  rotation?: number;
}

export type BackgroundNodeData = ImageNodeData | TextNodeData;

export interface BackgroundConfig {
  color?: string;
  nodes?: Node<BackgroundNodeData>[];
}

export interface Settings {
  title?: string;
  background?: BackgroundConfig;
  language?: string;
}

export interface EdgeConfig {
  animated?: boolean;
  color?: string;
  width?: number;
  type?: string;
}

export interface RoadmapData {
  nodes?: Node<NodeData>[];
  edges?: Edge[];
  settings: Settings;
  version: number;
}

export interface RoadmapState {
  nodes: Record<string, { state: string }>;
  x: number;
  y: number;
  zoom: number;
}

export type Orientation = "horizontal" | "vertical";

export type HelperLine = {
  // Used to filter out helper lines corresponding to the node being dragged
  node: Node;
  // We use it to check that the helper line is within the viewport.
  nodeBox: Box;
  // 0 for horizontal, 1 for vertical
  orientation: Orientation;
  // If orientation is 'horizontal', `position` holds the Y coordinate of the helper line.
  // (Might correspond to the top or bottom position of a node, or other anchors).
  // If orientation is 'vertical', `position` holds the X coordinate of the helper line.
  position: number;
  // Optional color for the helper line
  color?: string;
  anchorName: string;
};
