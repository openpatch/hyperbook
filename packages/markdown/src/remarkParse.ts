// @ts-nocheck
import { Root } from "mdast";
import { fromMarkdown } from "mdast-util-from-markdown";
import { Plugin } from "unified";
import micromarkSubSup from "./micromarkSubSup";
import mdastSubSup from "./mdastSubSup";

const remarkParse: Plugin<[], string, Root> = function () {
  const self = this;

  self.parser = function (doc) {
    const extensions = self.data("micromarkExtensions") || [];
    const mdastExtensions = self.data("fromMarkdownExtensions") || [];

    return fromMarkdown(doc, {
      ...self.data("settings"),
      extensions: [micromarkSubSup, ...extensions],
      mdastExtensions: [mdastSubSup, ...mdastExtensions],
    });
  };
};

export default remarkParse;
