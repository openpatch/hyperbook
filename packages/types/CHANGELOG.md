# @hyperbook/types

## 0.0.1

### Patch Changes

- [#316](https://github.com/openpatch/hyperbook/pull/316) [`cedb551`](https://github.com/openpatch/hyperbook/commit/cedb55191fd025b5a214df406a53cbab5d1b1bc1) Thanks [@mikebarkmin](https://github.com/mikebarkmin)! - Extract Hyperbook types into a separate package
