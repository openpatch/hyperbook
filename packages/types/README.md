# @hyperbook/types

## Installation

```sh
yarn add @hyperbook/types
# or
npm i @hyperbook/types
```
