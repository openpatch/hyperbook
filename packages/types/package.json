{
  "name": "@hyperbook/types",
  "version": "0.5.0",
  "author": "Mike Barkmin",
  "homepage": "https://github.com/openpatch/hyperbook#readme",
  "license": "MIT",
  "main": "dist/index.cjs.js",
  "module": "dist/index.esm.mjs",
  "types": "dist/index.d.ts",
  "typings": "dist/index.d.ts",
  "sideEffects": false,
  "exports": {
    ".": {
      "require": "./dist/index.cjs.js",
      "default": "./dist/index.esm.mjs"
    }
  },
  "files": [
    "dist"
  ],
  "publishConfig": {
    "access": "public"
  },
  "repository": {
    "type": "git",
    "url": "git+https://github.com/openpatch/hyperbook.git",
    "directory": "packages/types"
  },
  "bugs": {
    "url": "https://github.com/openpatch/hyperbook/issues"
  },
  "scripts": {
    "prebuild": "rimraf dist",
    "version": "pnpm build",
    "lint": "tsc --noEmit",
    "build": "pnpm build:pkg && pnpm build:types",
    "build:pkg": "node ../../scripts/build.mjs",
    "build:types": "tsc --project tsconfig.build.json --declaration --emitDeclarationOnly"
  },
  "dependencies": {}
}
