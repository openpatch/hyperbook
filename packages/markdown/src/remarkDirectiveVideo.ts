// Register directive nodes in mdast:
/// <reference types="mdast-util-directive" />
//
import { HyperbookContext } from "@hyperbook/types";
import { Comment, Element, Text } from "hast";
import { Root } from "mdast";
import { Raw } from "mdast-util-to-hast";
import { visit } from "unist-util-visit";
import { VFile } from "vfile";
import {
  expectLeafDirective,
  isDirective,
  registerDirective,
} from "./remarkHelper";

export default (ctx: HyperbookContext) => () => {
  const name = "video";
  return (tree: Root, file: VFile) => {
    visit(tree, function (node) {
      if (isDirective(node)) {
        if (node.name !== name) return;

        const data = node.data || (node.data = {});
        const attributes = node.attributes || {};
        const { src, title, author, poster } = attributes;

        expectLeafDirective(node, file, name);
        registerDirective(file, name, [], ["style.css"], []);

        const information: (Text | Comment | Element | Raw)[] = [];
        if (title) {
          information.push({
            type: "element",
            tagName: "span",
            properties: {
              class: "title",
            },
            children: [
              {
                type: "text",
                value: title,
              },
            ],
          });
        }
        if (title && author) {
          information.push({
            type: "element",
            tagName: "span",
            properties: {
              class: "spacer",
            },
            children: [
              {
                type: "text",
                value: "-",
              },
            ],
          });
        }
        if (author) {
          information.push({
            type: "element",
            tagName: "span",
            properties: {
              class: "author",
            },
            children: [
              {
                type: "text",
                value: author,
              },
            ],
          });
        }

        data.hName = "div";
        data.hProperties = {
          class: "directive-video",
        };
        data.hChildren = [
          {
            type: "element",
            tagName: "div",
            properties: {
              class: "player",
            },
            children: [
              {
                type: "element",
                tagName: "video",
                properties: {
                  controls: true,
                  src: src
                    ? ctx.makeUrl(
                        src,
                        "public",
                        ctx.navigation.current || undefined,
                      )
                    : undefined,
                  poster: poster
                    ? ctx.makeUrl(
                        poster,
                        "public",
                        ctx.navigation.current || undefined,
                      )
                    : undefined,
                  width: "100%",
                },
                children: [
                  {
                    type: "text",
                    value: "Your browser does not support videos.",
                  },
                ],
              },
            ],
          },
          {
            type: "element",
            tagName: "div",
            properties: {
              class: "information",
            },
            children: information,
          },
        ];
      }
    });
  };
};
