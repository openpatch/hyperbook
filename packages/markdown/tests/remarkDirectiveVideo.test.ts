import { HyperbookContext } from "@hyperbook/types/dist";
import { describe, expect, it } from "vitest";
import rehypeStringify from "rehype-stringify";
import remarkToRehype from "remark-rehype";
import { unified, PluggableList } from "unified";
import remarkDirective from "remark-directive";
import remarkDirectiveRehype from "remark-directive-rehype";
import rehypeFormat from "rehype-format";
import { ctx } from "./mock";
import remarkDirectiveVideo from "../src/remarkDirectiveVideo";
import remarkParse from "../src/remarkParse";

export const toHtml = (md: string, ctx: HyperbookContext) => {
  const remarkPlugins: PluggableList = [
    remarkDirective,
    remarkDirectiveRehype,
    remarkDirectiveVideo(ctx),
  ];

  return unified()
    .use(remarkParse)
    .use(remarkPlugins)
    .use(remarkToRehype)
    .use(rehypeFormat)
    .use(rehypeStringify, {
      allowDangerousCharacters: true,
      allowDangerousHtml: true,
    })
    .processSync(md);
};

describe("remarkDirectiveVideo", () => {
  it("should transform with author and title", async () => {
    expect(
      toHtml(
        `::video{src="/clouds.mp4" poster="/clouds.jpg" title="Morgenwolken" author="Natureclip (CC-BY 3.0)"}
`,
        ctx,
      ).value,
    ).toMatchSnapshot();
  });
  it("should transform with author", async () => {
    expect(
      toHtml(
        `::video{src="/clouds.mp4" poster="/clouds.jpg" author="Natureclip (CC-BY 3.0)"}
`,
        ctx,
      ).value,
    ).toMatchSnapshot();
  });
  it("should transform with title", async () => {
    expect(
      toHtml(
        `::video{src="/clouds.mp4" poster="/clouds.jpg" title="Morgenwolken"}
`,
        ctx,
      ).value,
    ).toMatchSnapshot();
  });
  it("should transform with missing src", async () => {
    expect(
      toHtml(
        `::video{poster="/clouds.jpg" title="Morgenwolken"}
`,
        ctx,
      ).value,
    ).toMatchSnapshot();
  });
  it("should transform with missing src", async () => {
    expect(
      toHtml(
        `::video{poster="/clouds.jpg" title="Morgenwolken"}
`,
        ctx,
      ).value,
    ).toMatchSnapshot();
  });
  it("should fail on wrong column count", async () => {
    expect(
      toHtml(
        `:video{poster="/clouds.jpg" title="Morgenwolken"}
`,
        ctx,
      ).messages.length,
    ).toBe(1);
  });
  it("should register directives", async () => {
    expect(
      toHtml(`::video{poster="/clouds.jpg" title="Morgenwolken"}`, ctx).data
        .directives?.["video"],
    ).toBeDefined();
  });
});
