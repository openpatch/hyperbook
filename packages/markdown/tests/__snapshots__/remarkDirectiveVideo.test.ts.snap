// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`remarkDirectiveVideo > should transform with author 1`] = `
"
<div class="directive-video">
  <div class="player">
    <video controls src="/public/clouds.mp4" poster="/public/clouds.jpg" width="100%">Your browser does not support videos.</video>
  </div>
  <div class="information"><span class="author">Natureclip (CC-BY 3.0)</span></div>
</div>
"
`;

exports[`remarkDirectiveVideo > should transform with author and title 1`] = `
"
<div class="directive-video">
  <div class="player">
    <video controls src="/public/clouds.mp4" poster="/public/clouds.jpg" width="100%">Your browser does not support videos.</video>
  </div>
  <div class="information"><span class="title">Morgenwolken</span><span class="spacer">-</span><span class="author">Natureclip (CC-BY 3.0)</span></div>
</div>
"
`;

exports[`remarkDirectiveVideo > should transform with missing src 1`] = `
"
<div class="directive-video">
  <div class="player">
    <video controls poster="/public/clouds.jpg" width="100%">Your browser does not support videos.</video>
  </div>
  <div class="information"><span class="title">Morgenwolken</span></div>
</div>
"
`;

exports[`remarkDirectiveVideo > should transform with missing src 2`] = `
"
<div class="directive-video">
  <div class="player">
    <video controls poster="/public/clouds.jpg" width="100%">Your browser does not support videos.</video>
  </div>
  <div class="information"><span class="title">Morgenwolken</span></div>
</div>
"
`;

exports[`remarkDirectiveVideo > should transform with title 1`] = `
"
<div class="directive-video">
  <div class="player">
    <video controls src="/public/clouds.mp4" poster="/public/clouds.jpg" width="100%">Your browser does not support videos.</video>
  </div>
  <div class="information"><span class="title">Morgenwolken</span></div>
</div>
"
`;
